src/thread/thread_d.hpp

```diff
@@ -524,10 +531,10 @@ namespace fc {
 
               next->_set_active_context( current );
               current->cur_task = next;
-              next->run();
+              fc::shared_ptr<task_base> next_ptr(next);
+              next_ptr->run();
               current->cur_task = 0;
-              next->_set_active_context(0);
-              next->release();
```

Comment:
Does this mean before the change, `next` never got freed? Would this PR fix https://github.com/bitshares/bitshares-core/issues/1006?

Reply:
> Does this mean before the change, `next` never got freed?
   Yes, in the case if thread::quit() was called the task "async thread::quit()" will be newer freed. I've added "fc::shared_ptr<task_base> next_ptr(next)" in this place just for so as not to explicitly call the "next->release();". 
   In the case "thread::quit()" an exception is raised from "my->start_next_fiber(true)" and everything will not called that is after (see run_next_task() {... next->run(); ... don't get here }.
In the ~thread_d I check if ready_context has a task and free the last.

Reply:
> [bitshares/bitshares-core#1006](https://github.com/bitshares/bitshares-core/issues/1006)?

Right now I have no answer. Necessary make additional research.

Reply:
> Does this mean before the change, next never got freed?

The call to `next->release()` will delete `*next` if this was the last reference to it. (`retainable` and its `retain`/`release` methods implement reference counting, used by `task_base` and `fc::shared_ptr`. It's a mess, don't ask. I'm working on eliminating this at least partially in https://github.com/bitshares/bitshares-core/issues/1584 .)

Reply:
I don't think `next->run()` will ever throw, so why change this here?

Reply:
> It's a mess, don't ask.

Yes, you are right absolutely. )

Reply:
> I don't think `next->run()` will ever throw, so why change this here?

Yes, this change is not necessary, as I wrote before, just  for so as not to explicitly call `release`.